### ISSUE TITLE
external storage with several users never sync

### ISSUE THREAD

**Issue by DeepDiver1975** [owncloud/core#13456]
reported by @kaneo-gmbh in https://github.com/owncloud/client/issues/2714

Hello,

we encounterd a problem while using the external storage feature of owncloud.

We have a smb-folder configured with passing the ownCloud user/password for smb authentication. The folder is accessible read/write for several users.

If one user is adding or modifying a file, the file will sync as desired from the users desktop-client or the users web-interface to the server. This step works reliably.

Unfortunately none of the other users that have access to this smb-folder will ever receive this file while using the desktop client. The desktop client seems never to notice, that files have been changed on the external storage. When the user logs on to the web-interface, the desktop client suddenly notices the changes and starts to sync.

Our workaround is, to log on in parallel to the web-interface.

**Comment by khlem** [comment]
Hello,

I have the same bug.

Windows Client : 1.7.1 (build 4382)
Owncloud Server : 7.0.4.2
OS : Debian 7
Apache : 2.2.22
PHP : 5.4.36

This problem is present for all Client sync (MAC, Linux, Windows) when the external smb module is activated. I have one external folder hosted on a Windows 2008 R2.

The only solution for me is to put new files/folder for force the smb discovery.

Do you have any information about this bug ?

Thx for your help.

Regards,

Clément.

**Comment by DeepDiver1975** [comment]
> Do you have any information about this bug ?

we will address this issue within OC8.1

**Comment by khlem** [comment]
Ok thx for your reply :)

**Comment by PVince81** [comment]
This is a general issue with external storages that remote changes cannot be reliably detected, see https://github.com/owncloud/core/issues/11797

When you change a file remotely on the SMB drive, does the mtime of the parent folders automatically change on the SMB server ? Because OC relies on that to detect changes.

**Comment by khlem** [comment]
In fact i mounted a SMB CIFS because my OC is fully integrated in my Active Directory 2008.

All data are stored in NTFS shared folder Windows 2008R2 with different right.

I can see new file/folder through the Web but nothing with Client sync.

When you change a file remotely on the SMB drive, does the mtime of the parent folders automatically change on the SMB server ?

Nop the mtime changes only for the file not for the parent folders.
For exemple : i can create a folder in Windows (creation date 02/01/2015) but when i add a new file or a new folder the creation date of the parent folder does not change.

Question : is it possible to force the discovery of folder in the Client sync ?
We could have a button (force discovery) or a timer to check the folder synchronised.

**Comment by khlem** [comment]
Hello,

Just for update.

I tried to update the "folder time" in parent folder but the issue persists.

So, i tried a different approach.

When i mount an external Windows shared folder with smb client and i mount it (local folder) in Owncloud, the Client sync is fully functionnal.

Regards,

Clément.

**Comment by khlem** [comment]
Hello,

I have done several test and i know why the Client does not sync with new files added in SMB shared (auth CIF/OC).

I use SMB CIF/OC for mount the Windows shared folder. When i am connected in OC website, or when i add manually a new file with the sync client, the new files are detected.
The issue appears when these above actions has not done.

I execute the "occ scan files" command in the server and i see the external SMB is not scanned, i reproduce this action with an external SMB CIF (without OC), the external SMB is totally scanned.
In fact it is logical because the auth is done with a login/password write in OC which is not the case with SMB CIF/OC (the auth is ask when the user is logged at the OC website).

I am not a developer but i think the best way is to add the occ scan command in the client sync and compare the files already scanned (compare cache ?).
When the sync client is connected to OC through the sync client the auth is already done.

If i can help you....

Regards,

Clément.

**Comment by directict** [comment]
i really would like to see this fixed in 8.1.. we now only use SMB trough WEB and tell our customer to NOT sync trough client!

**Comment by alefattorini** [comment]
Same problem here, do you have any workround?

**Comment by PVince81** [comment]
Can you guys post your "oc_storages" table ?

Are all your users using the same SMB/CIFS mount with the SAME credentials ? Or did you use the "SMB OC-Login" or the "$user" variable ?

If the former (same credentials), the etag propagation should work properly and updates should be detected.
In the latter case (separate credentials), every user will get their own "oc_storages" entry (and also own file cache). The storages cannot be linked together because OC has no way of knowing that it is the same share / remote folder, so the only mechanism that can be used to detect changes is detecting remote changes, as if a file was changed outside of ownCloud. In this specific case, the detection is done based on the mtime, but is insufficient because it cannot detect changes deep in the folder structures (see https://github.com/owncloud/core/issues/11797).

So far the only solution for this is to setup a cron job that runs `./occ files:scan --all` (or for specific users) to make sure a full rescan of the remote storage is done, in order to detect changes.

**Comment by PVince81** [comment]
@Xenopathic pointed out to me that the former case also won't propagate etags past external storage mount point boundaries like shares do. The former case (shared storage entry) should be fixable by making the etag propagation continue up to every user's root. (CC @icewind1991)

**Comment by khlem** [comment]
Hello,

Each user use separate credentials and the external storage is configured with SMB/CIFS OC.

The cmd : ./occ files:scan --all  scan only the local directory not the shared directory (the same observation with specific user).

oc_storages :

smb::XXXX\cloud@archi-data//data//
49
smb::XXXX\cloud@archi-data//syncloud//
50
smb::XXXX\test@archi-data///
45
smb::XXXX\test@archi-data//data//
23
smb::XXXX\test2@archi-data//data//
25
smb::XXXX\test2@archi-data//syncloud//
47

Thx for your help !

**Comment by PVince81** [comment]
@khlem I believe this issue was fixed in 8.1, that it would not go into mount points. Which version are you using ?

**Comment by khlem** [comment]
My production version is : 8.0.6 (stable)
My test version is : 8.1.2
I have the same problem to detect the new files with the Owncloud Client.

The Oc_storages informations provided of the 8.0.6.

**Comment by caguiar** [comment]
For me:
Each user uses separate credentials and the external storage is configured with SMB/CIFS OC.

**Comment by khlem** [comment]
New test :
OC version 8.1.3 (my test version)
External drive : SMB/CIFS OC (folder shared in Windows 2008R2)
Ldap integration

OC_storages :

home::E1F6E487-4501-4717-B214-62769493E5F1
22
local::/var/www/owncloud/data/
24
smb::xxxx@archi-data//data//
23

Modification of icewind : smb_oc (add the domain) and server.php (fr_FR_utf8).

I try with the cmd ./occ files:scan (LDAP UID - specific user).
Folders/files aren't update.

**Comment by PVince81** [comment]
@khlem does the files:scan now enter all folders ?

**Comment by khlem** [comment]
Nop

root@owncloud:/var/www/owncloud# sudo -u www-data php occ files:scan E1F6E487-4501-4717-B214-62769493E5F1

Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/cache
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/files
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files/ownCloudClientManual.pdf
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files/README.txt
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files/ownCloudServerAdminManual.pdf
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files/ownCloudUserManual.pdf
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/cache
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/versions
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/files
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/keys
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/versions
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/files
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/files/Screen Shot 2015-09-17 at 17.34.04.png.d1442847610
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/files/Screen Shot 2015-09-17 at 17.33.55.png.d1442847608
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/files/Screen Shot 2015-09-17 at 17.44.43.png.d1442847629
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/files/test.txt.d1442846073
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/files_trashbin/keys
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111495
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/70732
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111494
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111471
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111495
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111495/716-883-max.png
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111495/36-36.png
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/70732
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/70732/36-36.png
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/70732/2048-2048-max.png
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111494
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111494/36-36.png
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111494/197-111-max.png
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111471
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111471/72-72.png
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/thumbnails/111471/2048-2048-max.png

**Comment by PVince81** [comment]
Hmmm... can you try with `occ files:scan --all` or `occ files:scan --path /E1F6E487-4501-4717-B214-62769493E5F1/files/` ?

**Comment by PVince81** [comment]
@icewind1991 it is likely that mounting external storage during scan doesn't work properly with LDAP ?

**Comment by khlem** [comment]
root@owncloud:/var/www/owncloud# sudo -u www-data php occ files:scan --path /E1F6E487-4501-4717-B214-62769493E5F1/files
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files
Scanning folder /E1F6E487-4501-4717-B214-62769493E5F1/files
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files/ownCloudClientManual.pdf
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files/README.txt
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files/ownCloudServerAdminManual.pdf
Scanning file   /E1F6E487-4501-4717-B214-62769493E5F1/files/ownCloudUserManual.pdf

**Comment by icewind1991** [comment]
> it is likely that mounting external storage during scan doesn't work properly with LDAP ?

Not for any reason I'm aware of

**Comment by khlem** [comment]
it is likely that mounting external storage during scan doesn't work properly with LDAP ?
@PVince81 Exactly

**Comment by PVince81** [comment]
@khlem I did a local test on 8.1.3 and I am able to scan the mounts with LDAP:
```
vvortex:/srv/www/htdocs/owncloud # sudo -uwwwrun ./occ files:scan 3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/
Scanning folder /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/mount.json
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/thumbnails
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files_external
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/cache
Scanning folder /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/thumbnails
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/thumbnails/13
Scanning folder /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/thumbnails/13
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/thumbnails/13/39-39.png
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/thumbnails/13/2048-2048-max.png
Scanning folder /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files_external
Scanning folder /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/welcome.txt
Scanning folder /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/cache
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/sftp2/
Scanning folder /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/sftp2/
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/sftp2/P1040667.JPG
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/sftp2/P1040702.JPG
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/sftp2/P1040703.JPG
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/sftp/
Scanning folder /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/sftp/
Scanning file   /3927f8f4-f0a0-1034-9e3c-9f1dfdf8a5e8/files/sftp/x
```

The mount "sftp" is a system-wide SFTP mount and "sftp2" is a personal mount.

@jmaciasportela can you try this too with the WND app ?

**Comment by khlem** [comment]
@PVince81 My ldap is Windows 2008R2 (Microsoft) what is your ldap (openldap ?).

Maybe this issue is only with MS Ldap ?

**Comment by PVince81** [comment]
@khlem I have an openLDAP docker instance.

@blizzz @jmaciasportela can you help check if "occ files:scan" work with the above config ?

**Comment by khlem** [comment]
@PVince81 Thx to help us
If i can help you with some tests, no problem.

**Comment by caguiar** [comment]
here is my experience:

LDAP is from SAMBA v4
ownCloud Version: 8.1.1

--------
# sudo -u www-data ./occ files:scan 97C5B7E0-4380-41D7-B203-5EC3195AD4CC
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/
Scanning folder /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/files_external
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/files_trashbin
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/files
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/mount.json
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/files_versions
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/thumbnails
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/cache
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/lucene_index
Scanning folder /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/files_external
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/files_external/uploads
... etc etc ... until ...
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/lucene_index/_1.cfs
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/lucene_index/segments_3
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/lucene_index/read.lock.file
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/lucene_index/segments.gen
Scanning file   /97C5B7E0-4380-41D7-B203-5EC3195AD4CC/files/_BOLSAS_/
                                                   
[Icewind\SMB\Exception\AuthenticationException]  
------------
So I get this exception on the first folder (_BOLSAS_) which is external (shared SMB folder configured with SMB/CIFS OC).

**Comment by DeepDiver1975** [comment]
@karlitschek @cmonteroluque we have no solution at hand - neither for 8.2 nor anything less-risky which is backportable.

**Comment by khlem** [comment]
Why not use an agent who check the files modifications directly on the shared server ?
Example : The Net Framework >3.5 (Windows) via a filesystem watcher can handle the job.
In fact the agent scans and updates the cache of each account.
An API can be proposed and each developer can add his "file system watcher"

**Comment by PVince81** [comment]
The problem with watchers/daemons is that not everybody can set them up (ex: shared hosters). But it could be used as an optional addition to improve update detection.

**Comment by khlem** [comment]
I think the strength of Owncloud is the integration of many system and permit the use of LDAP.
The reinforcement of files update detection must be considered.
I hope see this update in the near future :).

**Comment by karlitschek** [comment]
definitely a tricky and complex question. Something for the future. 9.0

**Comment by khlem** [comment]
@karlitschek I hope for the Owncloud 9.0.
Thx for your answer.

**Comment by DeepDiver1975** [comment]
> definitely a tricky and complex question. Something for the future. 9.0

@karlitschek can you please adjust the milestone then? Only you and @cmonteroluque are allowed to - thx

**Comment by karlitschek** [comment]
done :-)

**Comment by DeepDiver1975** [comment]
> done :-)

THX

**Comment by PVince81** [comment]
Research still in progress for update detection: #11797

**Comment by PVince81** [comment]
Ok, do I understand it correctly that every user have their own access to the SMB mount ?

So it's not a single system-wide SMB mount in OC that is accessible by all users, using the "applicable" field in the web UI.
But it's every user has their own OC SMB mount from their personal pages, so every one get their own access with their own separate credentials.

This internally causes several storage entries to be used as they are all considered separate, especially due to potentially different permissions.

**Comment by khlem** [comment]
We have this issue with a single system SMB mount in OC for all users.
The change detection not function properly.
SMB Windows with NTFS gives permission to this or that folder and subfolder when the user add a new document or folder in the SMB system, OC doesn't detect the modification.
I tested with the last version of OC

**Comment by PVince81** [comment]
Closing in favor of https://github.com/owncloud/core/issues/11797
